File: Lab_5/src/main/scala/lab5/util/ArrayUtil.scala
package lab5.util

import java.util.Random

object ArrayUtil:
  extension [T](arr: Array[T])
    def swap(idx1: Int, idx2: Int): Unit =
      val tmp = arr(idx1)
      arr(idx1) = arr(idx2)
      arr(idx2) = tmp

    def scramble(rng: Random): Unit =
      for i <- arr.indices.reverse.dropRight(1) do
        val idx = rng.nextInt(i + 1)
        arr.swap(idx, i)

File: Additional_1/src/main/java/additional1/util/ArrayUtil.java
package additional1.util;

import java.util.Random;

public class ArrayUtil {
    public static void swapElement(int[] a1, int[] a2, int idx) {
        int tmp = a1[idx];
        a1[idx] = a2[idx];
        a2[idx] = tmp;
    }

    public static void swapElement(int[] a, int idx1, int idx2) {
        int tmp = a[idx1];
        a[idx1] = a[idx2];
        a[idx2] = tmp;
    }

    public static void scrambleArray(int[] arr, Random random) {
        for (int i = arr.length - 1; i > 0; i--) {
            int index = random.nextInt(i + 1);
            swapElement(arr, index, i);
        }
    }

    public static void reverse(int[] arr) {
        for (int i = 0; i < arr.length / 2; i++) swapElement(arr, i, arr.length - i - 1);
    }
}
